### PR TITLE
feat: show demo on demand in labs

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

~The UX is barely there~, but you can now click on the ~text~ button "Show Demo" in the instructions of a lab and the solution should appear in the modal. Thanks to Huyen for smartening up the UX.

### CODE REVIEW

File: client/src/templates/Challenges/components/tool-panel.tsx [hunk 1]
```diff
@@ -34,6 +34,7 @@ const mapDispatchToProps = (dispatch: Dispatch) =>
       openHelpModal: () => openModal('help'),
       openVideoModal: () => openModal('video'),
       openResetModal: () => openModal('reset'),
+      // openSolutionPreviewModal: () => openModal('');
```
Comment (moT01) [inline client/src/templates/Challenges/components/tool-panel.tsx:37]:
```suggestion
```

**Review by moT01 — APPROVED** [review]
This is awesome @ojeytonwilliams 🎉 

The modal that shows up for the labs says "Here's a preview of what you will build" at the top. I think we should change that to something else - maybe "Here's an example of a project that fulfills the user stories." And it says "Start Coding!" at the bottom. That doesn't feel quite right either, maybe just "Code". Also we probably want something in the instruction panel around that button, "Build a project that is functionally similar to this: button"

I don't think any of that needs to be added here, this adds the main functionality and isn't visible anywhere except the new labs - we can make subsequent PR's later.

Thinking about this now, I wonder if we could use that portal for all these. I would think so. That would probably be better. We don't need to change it now, but at some point, maybe we can move to that.

Edit: This seems to have broken the project view in the settings/timeline/portfolio. When I click "view project" on one of those, the preview is blank.

Edit: Actually, it seems to be working. I'm not sure why it wasn't showing up right away. I made some changes, and it showed up - then I reverted them to what's on this PR and it worked. Do you want to give that a quick check @Oliver? Just submit one of the projects and make sure the preview shows up in the modal when you click "view project" on the settings page.

**Comment by ojeytonwilliams** [comment]
>  Just submit one of the projects and make sure the preview shows up in the modal when you click "view project" on the settings page.

Yep, it seems fine. It showed up on the timeline, too.

**Comment by ojeytonwilliams** [comment]
> The modal that shows up for the labs says "Here's a preview of what you will build" at the top. I think we should change that to something else - maybe "Here's an example of a project that fulfills the user stories." And it says "Start Coding!" at the bottom. That doesn't feel quite right either, maybe just "Code". Also we probably want something in the instruction panel around that button, "Build a project that is functionally similar to this: button"

I like the sound of all that. Definitely "Nice to have", though.

**Comment by zairahira** [comment]
I would agree with Tom and we should change the label "Here's a preview of what you will build" at the top to "Here's an example of a project that fulfills the user stories."
 
Also is this change specifically for labs?

**Comment by ojeytonwilliams** [comment]
@zairahira the change to the label makes sense, but I'd prefer to do that in a followup PR.

> Also is this change specifically for labs?

It's intended to be used by labs, but it's quite general. If you want to show the demo (i.e. the solution to the last challenge), then you can do that on any challenge by setting the `demoType` property. The value `workshop` vs `labs` determines if it will be shown automatically or if we should show a button to show it.

Given that, I think it makes sense to change the values from `workshop` and `labs` to `onLoad` and `onClick` respectively. That better reflects what they are, rather how they're used. Does that make sense?

**Comment by ojeytonwilliams** [comment]
I went ahead and changed the names. Happy to revert/change if people hate them.

**Comment by zairahira** [comment]
I tested the `onLoad` and `onClick` functionality and it's working perfectly fine 👍️

**Comment by ShaunSHamilton** [comment]
So, the workshops are not supposed to show the preview on the first step?

**Comment by ojeytonwilliams** [comment]
It should be configurable. If you add `demoType` to the frontmatter of a challenge then it should show a demo. Either automatically, if you use `onLoad`, or when you click the button, if you use `onClick`.

So, it's up to whoever creates a lesson to decide which, if any, they want.

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍

**Comment by KorreterBrown** [comment]
#55569 <https://github.com/freeCodeCamp/freeCodeCamp/pull/55569>

Shaun Hamilton ***@***.***> 于2024年8月13日周二 21:07写道：

**Pull request merged**